initrd/bin/tpmr

```diff
@@ -0,0 +1,382 @@
+# no password was used to seal).
+tpm2_unseal() {
+	TRACE "Under /bin/tpmr:tpm2_unseal"
+	index="$1"
+	pcrl="$2"
+	sealed_size="$3"
+	file="$4"
+	pass="$5"
+
+	# TPM2 doesn't care about sealed_size, only TPM1 needs that.  We don't
+	# have to separately read the sealed file on TPM2.
+
+	# Pad with up to 6 zeros, i.e. '0x81000001', '0x81001234', etc.
+	handle="$(printf "0x81%6s" "$index" | tr ' ' 0)"
+
+	DEBUG "tpm2_unseal: handle=$handle pcrl=$pcrl file=$file pass=$([ "$pass" ] && echo "<yes>" || echo "<no>")"
+
+	# If we don't have the primary handle (TPM hasn't been reset), tpm2 will
+	# print nonsense error messages about an unexpected handle value.  We
+	# can't do anything without a primary handle.
+	if [ ! -f "/tmp/$PRIMARY_HANDLE_FILE" ]; then
+		DEBUG "tpm2_unseal: No primary handle, cannot attempt to unseal"
+		exit 1
+	fi
+
+	POLICY_SESSION=/tmp/unsealfile_policy.session
+	rm -f "$POLICY_SESSION"
+	tpm2 startauthsession -Q -g sha256 -S "$POLICY_SESSION" --policy-session
+	trap "cleanup_session '$POLICY_SESSION'" EXIT
+	# Check the PCR policy
+	tpm2 policypcr -Q -l "$pcrl" -S "$POLICY_SESSION"
+	UNSEAL_PASS_SUFFIX=""
+
+	HMAC_SESSION=/tmp/unsealfile_hmac.session
+	tpm2 startauthsession -g sha256 -c "/tmp/$PRIMARY_HANDLE_FILE" -S "$HMAC_SESSION" --hmac-session
+	trap "cleanup_session '$POLICY_SESSION'" EXIT
+
+	if [ "$pass" ]; then
+		# Add the object authorization policy (the actual password is
+		# provided later, but we must include this so the policy we
+		# attempt to use is correct).
+		tpm2 policypassword -Q -S "$POLICY_SESSION"
+		# When unsealing, include the password with the auth session
+		UNSEAL_PASS_SUFFIX="+hex:$(echo -n "$pass" | sha256sum | cut -d ' ' -f 1)"
+	fi
+
+	tpm2 unseal -Q -c "$handle" -p "session:$POLICY_SESSION$UNSEAL_PASS_SUFFIX" -S "$HMAC_SESSION" > "$file"
+}
+
+tpm2_reset() {
+	TRACE "Under /bin/tpmr:tpm2_reset"
+	key_password="$1"
+	mkdir -p "$SECRET_DIR"
+	tpm2 clear -c platform || warn "Unable to clear TPM on platform hierarchy!"
+	tpm2 changeauth -c owner "$key_password"
+	tpm2 createprimary -C owner -g sha256 -G "${CONFIG_PRIMARY_KEY_TYPE:-rsa}" -c "$SECRET_DIR/primary.ctx" -P "$key_password"
+	tpm2 evictcontrol -C owner -c "$SECRET_DIR/primary.ctx" "$PRIMARY_HANDLE" -P "$key_password"
+	shred -u "$SECRET_DIR/primary.ctx"
+	tpm2_startsession
+}
+
+# Perform final cleanup before boot and lock the platform heirarchy.
+tpm2_kexec_finalize() {
+	# Flush sessions and transient objects
+	tpm2 flushcontext -Q --transient-object \
+		|| warn "tpm2_flushcontext: unable to flush transient handles"
+	tpm2 flushcontext -Q --loaded-session \
+		|| warn "tpm2_flushcontext: unable to flush sessions"
+	tpm2 flushcontext -Q --saved-session \
+		|| warn "tpm2_flushcontext: unable to flush saved session"
+
+	# Add a random passphrase to platform hierarchy to prevent TPM2 from
+	# being cleared in the OS.
+	# This passphrase is only effective before the next boot.
+	echo "Locking platform hierarchy..."
```

Comment:
@JonathonHall-Purism :  "Locking TPM2 platform hierarchy..." ?

Reply:
Sure, will change it